tests/consensus_test.rs

```diff
@@ -501,348 +576,385 @@ impl TestNetwork {
         assert!(
-            network.nodes[i].num_shard_chunks().await >= 3,
-            "Node {} should have confirmed blocks",
-            i
+            node.num_shard_chunks() >= num_chunks,
+            "Node (read={}, idx={}) should have confirmed at least {} shard chunks, but has {}",
+            is_read,
+            index,
+            num_chunks,
+            node.num_shard_chunks()
         );
+    });
+}
 
+fn assert_network_has_messages(network: &TestNetwork, num_messages: usize) {
+    on_all_nodes(network, |node, is_read, index| {
         assert!(
-            network.nodes[i].total_messages().await > 0,
-            "Node {} should have messages",
-            i
+            node.total_messages() >= num_messages,
+            "Node (read={}, idx={}) should have at least {} messages, but has {}",
+            is_read,
+            index,
+            num_messages,
+            node.total_messages()
         );
-    }
+    });
 }
 
-async fn wait_for_blocks(node: &NodeForTest, num_blocks: usize) {
-    let timeout = tokio::time::Duration::from_secs(10);
-    let start = tokio::time::Instant::now();
-    let mut timer = time::interval(tokio::time::Duration::from_millis(10));
-    let existing_blocks = node.num_blocks().await;
-
-    loop {
-        let _ = timer.tick().await;
-        if node.num_blocks().await >= existing_blocks + num_blocks {
-            break;
-        }
-        if start.elapsed() > timeout {
-            break;
-        }
-    }
-
-    assert!(
-        node.num_blocks().await >= existing_blocks + num_blocks,
-        "Node should have confirmed blocks"
-    );
-    assert!(
-        node.num_shard_chunks().await >= existing_blocks + num_blocks,
-        "Node should have confirmed shard chunks"
-    );
+fn assert_network_has_cast(network: &TestNetwork, fid: u64, hash: Vec<u8>) {
+    on_all_nodes(network, |node, is_read, index| {
+        assert!(
+            node.cast_added(fid, hash.clone()).is_some(),
+            "Node (read={}, idx={}) should have cast message with fid {} and hash {}, but it was not found",
+            is_read,
+            index,
+            fid,
+            hex::encode(hash.clone())
+        );
+    });
 }
 
 #[tokio::test]
 #[serial]
-async fn test_basic_sync() {
+async fn test_basic_consensus() {
     let env_filter = EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("warn"));
     let _ = tracing_subscriber::fmt()
         .with_env_filter(env_filter)
         .try_init();
 
-    // Set up shard and validators
-    let num_shards = 1;
+    let num_shards = 2;
+    let mut network = TestNetwork::create(3, num_shards).await;
+    network.start_validators().await;
+
+    network.register_and_wait_for_fid(1000).await;
+    let cast = network
+        .send_and_wait_for_cast(1000, "Hello, world")
+        .await
+        .unwrap();
 
-    let mut network = TestNetwork::create(4, num_shards, 3220).await;
+    assert_network_has_num_blocks(&network, 1);
```

Comment:
Were we previously asserting 3 here? Did the test fail with 3?

Reply:
I've adjusted this test to have all the nodes sync at the next block height before running assertions, and then ensuring that they've all reached that block. 

I think one of the issues that caused flakiness in the previous tests was that the nodes were all potentially at different heights, and therefore we relied on sleeping/producing blocks in order for them to converge. The approach I'm favoring now is to be more explicit about sync through these of these `wait*` methods to ensure nodes have converged.